### PR TITLE
store/tikv: copIterator should Close gracefully

### PR DESCRIPTION
**Pull request by tiancaiamao**
Fix a case that make caller of `copIterator.Next()` hang forever.

It's found when I try to do "select xxx order by desc limit" on a big table, several goroutine seems to be leak:

```
(pprof) top50 -cum
101 of 103 total (98.06%)
Showing top 50 nodes out of 83 (cum >= 1)
flat flat% sum% cum cum%
0 0% 0% 103 100% runtime.goexit /media/genius/OS/project/go/src/runtime/asm_amd64.s:2197
101 98.06% 98.06% 101 98.06% runtime.gopark /media/genius/OS/project/go/src/runtime/proc.go:271
0 0% 98.06% 66 64.08% runtime.chanrecv /media/genius/OS/project/go/src/runtime/chan.go:513
0 0% 98.06% 66 64.08% runtime.goparkunlock /media/genius/OS/project/go/src/runtime/proc.go:277
0 0% 98.06% 64 62.14% runtime.chanrecv2 /media/genius/OS/project/go/src/runtime/chan.go:400
0 0% 98.06% 58 56.31% github.com/pingcap/tidb/distsql.(*selectResult).fetch /media/genius/OS/project/src/github.com/pingcap/tidb/distsql/distsql.go:86
0 0% 98.06% 58 56.31% github.com/pingcap/tidb/store/tikv.(*copIterator).Next /media/genius/OS/project/src/github.com/pingcap/tidb/store/tikv/coprocessor.go:410
0 0% 98.06% 28 27.18% runtime.selectgo /media/genius/OS/project/go/src/runtime/select.go:238
0 0% 98.06% 28 27.18% runtime.selectgoImpl /media/genius/OS/project/go/src/runtime/select.go:423
0 0% 98.06% 7 6.80% net.(*pollDesc).wait /media/genius/OS/project/go/src/net/fd_poll_runtime.go:75
```
Those goroutines block at line coprocessor.go:410, when calling `copIterator.Next()` at line distsql.go:86

### CODE REVIEW

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -320,36 +318,45 @@ func (it *copIterator) work(ctx goctx.Context, taskCh <-chan *copTask) {
...
+			close(task.respChan)
 		}
 
-		for _, resp := range resps {
-			select {
-			case ch <- resp:
-			case <-ctx.Done():
-				return
-			case <-it.finished:
-				return
```
Comment (tiancaiamao) [inline store/tikv/coprocessor.go]: The bug is line 334 and here, it should close channel before return.

**Comment by tiancaiamao** [comment]
PTAL @zimulala @shenli

**Comment by ngaut** [comment]
Any tests?

**Comment by tiancaiamao** [comment]
I can reproduce it manually, but unit test is hard to write.

Here is how:
Prepare a table with 100w rows, select * from it, order by an index row, in descending order, limit 500 items.
When LimitExec get enough rows, it will close the executor, then XSelectIndexExec will be closed,
and also the copIterator. But a background goroutine (created by `go Fetch()`) is still calling copIterator's Next() to fetch handles,  that background would hang and leak.

To unit test it I can either:
Do the same thing as described above. I have to prepare a table big enough, so background goroutine will not finish fetching handles. However a table big enough is not feasible for unit test.
Or:
Test directly on copIterator, use a goroutine to call Next() and Close() the copIterator at the same time, test the goroutine never return. But it's hard to just make a copIterator and build such case.

@ngaut

**Comment by ngaut** [comment]
Could you implement the Fetch interface and wrap an inner origin object inside it? so you can simulate the slow fetch process by sleep for a while.

**Comment by tiancaiamao** [comment]
Seems this just happen on a revise branch, not master.
So this is not a serious problem, as long as Next() is never called after Close().

**Pull request closed without merging**